src/components/ec-file-upload/ec-file-upload.vue

```diff
@@ -0,0 +1,125 @@
+<template>
+  <div :data-test="$attrs['data-test'] ? `${$attrs['data-test']} ec-file-upload` : 'ec-file-upload'">
+    <label
+      v-if="label || note"
+      class="ec-file-upload__label"
+    >
+      <span
+        v-if="label"
+        data-test="ec-file-upload__label-text"
+        class="ec-file-upload__label-text"
+      >{{ label }}</span>
+
+      <span
+        v-if="note"
+        data-test="ec-file-upload__label-note"
+        class="ec-file-upload__label-note"
+      >{{ note }}</span>
+    </label>
+
+    <ec-file-dropzone
+      data-test="ec-file-upload__dropzone"
+      :is-disabled="isDisabled"
+      @change="onChange"
+    >
+      <slot
+        name="title"
+        slot="title"
+      />
+
+      <slot
+        name="subtitle"
+        slot="subtitle"
+      />
+    </ec-file-dropzone>
+
+    <div
+      data-test="ec-file-upload__help-text"
+      class="ec-file-upload__help-text"
+    >
+      <slot name="help-text" />
+    </div>
+
+    <ec-file-list
+      data-test="ec-file-upload__list"
+      :items="value"
+      :is-delete-disabled="isDisabled"
+      @delete="onDelete"
+    />
+  </div>
+</template>
+
+<script>
+import EcFileDropzone from '../ec-file-dropzone';
+import EcFileList from '../ec-file-list';
+
+export default {
+  name: 'EcFileUpload',
+  components: {
+    EcFileDropzone,
+    EcFileList,
+  },
+  model: {
+    prop: 'value',
+    event: 'change',
+  },
+  props: {
+    value: {
+      type: Array,
+      default: () => ([]),
+    },
+    label: {
+      type: String,
+    },
+    note: {
+      type: String,
+    },
+    isDisabled: {
+      type: Boolean,
+    },
+  },
+  methods: {
+    onChange(newFiles) {
+      let updatedFileList = [...this.value];
+
+      for (const newFile of newFiles) {
```

Comment:
I think could be enough only with having inside this `for` loop the content of the `else` block, since there you already are removing any possible duplicated file, and push the new one after. I mean, just having:
```
for (const newFile of newFiles) {
  updatedFileList = updatedFileList.filter(prevFile => prevFile.name !== newFile.name);
  updatedFileList.push(newFile);
}
```

But going one step further, perhaps you can avoid having nested loops by doing something like this:
```
onChange(newFiles) {
  const newFilesNames = newFiles.map(file => file.name);
  const updatedFileList = this.value.filter(file => !newFilesNames.includes(file.name));
  updatedFileList.push(...newFiles);
  this.$emit('change', updatedFileList);
},
```

Reply:
Yes, this is super elegant, brilliant @jafcalvente!

Reply:
I disagree with this. We shouldn't filter users input without telling anything to the user - that's a bad UX. We should emit all files and let the parent to do the validation and display an error message about some files have been removed.

If you decide to keep it then we need to change the code to use Set() because `.include` inside of the `.filter` is O(n^2).
```js
const existingNames = new Set(this.value.map(file => file.name));
newFiles = newFiles.filter(file => !existingNames.has(file.name));
this.$emit('change', [this.value, newFiles]);
```

Reply:
For file uploads, it's not necessarily bad UX. I'd argue, that most file uploads behave that way, so you could expect such behaviour. In theory, we could emit a replace event and show an information toast message, but personally, I'd not do that except UX wants to have this in place.

Reply:
Also to consider is that whilst the filenames may be the same, the file may be different, so checking against file size and name may be beneficial if you would like to avoid the same file being added. I'd say the parent should do the validation as we may want to show a custom message of the sort such as 'You have already added XX file, do you want to continue?"